### PR TITLE
`extension` meta type

### PR DESCRIPTION
**Pull request by tal5**
## Description
Adds an `extension` meta type, which lets you add to existing meta docs (e.g. adding to a core meta doc in an impl).
The way it works is that every `MetaObject` now stores the raw values they had passed into `ApplyValue`, and then in extension meta's `PostCheck` (which runs for them before other meta types), they apply their own `RawValues` into their target meta, using their target meta's `RawValues` to add to it instead of overriding it.

## Additions

- `MetaDocs.META_TYPE_EXTENSION` - static `MetaType` for `extension` meta.
- `MetaDocs.Extensions` - `Dictionary` to hold all extension meta objects.
- `MetaExtension` - `MetaObject` for extension meta.
- `MetaObject.RawValues` - `Dictionary` to hold all of the raw values passed into `ApplyValue`.
- `MetaObject#CleanValue` - cleans a value before it's passed into `ApplyValue`; added so that `MetaExtension` can get the raw value.

## Changes

- Added `MetaDocs.META_TYPE_EXTENSION` to the `MetaDocs.META_TYPES` array.
- Added a getter for `MetaExtension` to `MetaDocs.MetaObjectGetters`.
- `MetaDocs#AllMetaObjects` now returns all extension meta objects as well.
- `MetaDocsLoader#LoadInObject` now puts the key & value in the new `MetaObject.RawValues` if they're valid.
- `MetaDocsLoader#LoadInObject` now uses the new `MetaObject#CleanValue`.

> [!NOTE]
Currently extension meta uses a simple `switch` case on the meta type to find their target meta - if I trusted my C# skills enough to do a large rework I'd probably have each `MetaType` store the `Directory` of all their meta objects and their getter, then replace `MetaDocs.MetaObjectGetters` with a map of name to meta type, probably auto-generated because `MetaType`s already have a name.
That way instead of having a switch we could just get the relevant meta type from there and access all of it's objects; would probably cleanup `@link` handling as well.
Although the `switch` is probably fine for now, I believe that's what `@link` does as well.

> [!NOTE]
The `include_existing` option on extension meta might not be ideal - it's currently meant for stuff like adding examples, where you don't want any previous value included; couldn't find any better way to do that without reworking stuff (e.g. having each meta object specify it's keys and certain info about them), so this might be good enough for now? although let me know if you have any other ideas/want(/want me) to attempt such a rework.

### CODE REVIEW

File: SharpDenizenTools/MetaObjects/MetaObject.cs [hunk 1]
```diff
@@ -49,6 +49,16 @@ public abstract class MetaObject
         /// <summary>The backing <see cref="MetaDocs"/> instance.</summary>
         public MetaDocs Meta;
 
+        /// <summary>The original raw values specified for the object</summary>
+        public Dictionary<string, string> RawValues = new();
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaObject.cs]: Unnecessary/irrelevant change

EDIT: This is meant to be re the whole separated `CleanValue` processing, I didn't realize when I first left this comment that `RawValues` in particular was being used for the extend logic. So comment is still relevant just on the wrong line.

File: SharpDenizenTools/MetaObjects/MetaExtension.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+﻿using FreneticUtilities.FreneticExtensions;
+using SharpDenizenTools.MetaHandlers;
+using System.Collections.Generic;
+
+namespace SharpDenizenTools.MetaObjects
+{
+    /// <summary>A extension to an existing meta object.</summary>
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaExtension.cs]: `An`

File: SharpDenizenTools/MetaObjects/MetaExtension.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        public override MetaType Type => MetaDocs.META_TYPE_EXTENSION;
+
+        /// <inheritdoc/>
+        public override string Name => ExtensionName;
+
+        /// <summary>The name of the extension.</summary>
+        public string ExtensionName;
+
+        /// <summary>The type and name of the object this extension meta extends.</summary>
+        public string Extend;
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaExtension.cs]: Why is one string holding two separate values? Separate em.

File: SharpDenizenTools/MetaObjects/MetaExtension.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        /// <inheritdoc/>
+        public override void PostCheck(MetaDocs docs)
+        {
+            Require(docs, Extend, ExtensionName);
+            RawValues.Remove("extend");
+            RawValues.Remove("name");
+            RawValues.Remove("include_existing");
+            string metaName;
+            string metaType = Extend.ToLowerFast().BeforeAndAfter(' ', out metaName);
+            MetaObject extended = metaType switch
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaExtension.cs]: This lookup should probably be a dict inside `MetaDocs`

File: SharpDenizenTools/MetaObjects/MetaExtension.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                "mechanism" => docs.Mechanisms.GetValueOrDefault(metaName, null),
+                "tag" => docs.Tags.GetValueOrDefault(metaName, null),
+                "objecttype" => docs.ObjectTypes.GetValueOrDefault(metaName, null),
+                "property" => docs.Properties.GetValueOrDefault(metaName, null),
+                "event" => docs.Events.GetValueOrDefault(metaName, null),
+                "action" => docs.Actions.GetValueOrDefault(metaName, null),
+                "language" => docs.Languages.GetValueOrDefault(metaName, null),
+                _ => null
+            };
+            if (extended == null)
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaExtension.cs]: `is null`

File: SharpDenizenTools/MetaObjects/MetaExtension.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            RawValues.Remove("extend");
+            RawValues.Remove("name");
+            RawValues.Remove("include_existing");
+            string metaName;
+            string metaType = Extend.ToLowerFast().BeforeAndAfter(' ', out metaName);
+            MetaObject extended = metaType switch
+            {
+                "command" => docs.Commands.GetValueOrDefault(metaName, null),
+                "mechanism" => docs.Mechanisms.GetValueOrDefault(metaName, null),
+                "tag" => docs.Tags.GetValueOrDefault(metaName, null),
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaExtension.cs]: `, null` is redundant on these

File: SharpDenizenTools/MetaObjects/MetaObject.cs [hunk 7]
```diff
@@ -49,6 +49,16 @@ public abstract class MetaObject
...
+
+        /// <summary>Cleans a meta value before applying it onto this meta object.</summary>
+        /// <param name="rawValue">The raw value to clean.</param>
+        public virtual string CleanValue(string rawValue)
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaObject.cs]: Unnecessary/irrelevant change

File: SharpDenizenTools/MetaObjects/MetaExtension.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            {
+                docs.LoadErrors.Add($"Extension '{ExtensionName}' has invalid target meta to extend '{Extend}'.");
+                return;
+            }
+            foreach ((string key, string value) in RawValues)
+            {
+                string newValue = IncludeExisting && extended.RawValues.TryGetValue(key, out string currentValue) ? currentValue + value : base.CleanValue(value);
+                if (!extended.ApplyValue(docs, key, newValue))
+                {
+                    docs.LoadErrors.Add($"Extension '{ExtensionName}' could not extend {extended.Type.Name.ToLowerFast()} meta '{extended.Name}', key/value pair '{key}' -> '{value}' is invalid.");
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaExtension.cs]: why the `ToLowerFast()`

File: SharpDenizenTools/MetaObjects/MetaObject.cs (hunk 1 above)
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaObject.cs]: RawValues cannot be accurately contained to a `Dictionary` like this, many keys (eg `Usages`, `Examples`, etc.) can validly have multiple entries under the same key

Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaObject.cs]: Changed to a `List<string>` value

File: SharpDenizenTools/MetaObjects/MetaExtension.cs (hunk 5 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaExtension.cs]: Fixed

File: SharpDenizenTools/MetaObjects/MetaExtension.cs (hunk 2 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaExtension.cs]: Fixed

File: SharpDenizenTools/MetaObjects/MetaExtension.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            return true;
+        }
+
+        /// <inheritdoc/>
+        public override void PostCheck(MetaDocs docs)
+        {
+            Require(docs, Extend, ExtensionName);
+            RawValues.Remove("extend");
+            RawValues.Remove("name");
+            RawValues.Remove("include_existing");
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaExtension.cs]: Hacking the RawValues set like this seems inappropriate, Extension type should have its own set of what values to edit, or at least only exclude inline (ie don't persist the change onto the underlying dict)

Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaExtension.cs]: Now excludes them in-line

File: SharpDenizenTools/MetaObjects/MetaObject.cs (hunk 7 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaObject.cs]: Removed

File: SharpDenizenTools/MetaObjects/MetaExtension.cs (hunk 8 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaExtension.cs]: Replaced

File: SharpDenizenTools/MetaObjects/MetaExtension.cs (hunk 4 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaExtension.cs]: WIP version of that rework pushed

File: SharpDenizenTools/MetaObjects/MetaExtension.cs (hunk 3 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaExtension.cs]: Separated into `target_type` and `target_name`, not 100% sure about these names so let me know if you have any suggestions

File: SharpDenizenTools/MetaHandlers/MetaDocs.cs [hunk 10]
```diff
@@ -115,6 +99,23 @@ public class MetaDocs
         /// <summary>Set of raw adjustable keys.</summary>
         public HashSet<string> RawAdjustables = [];
 
+        /// <inheritdoc/>
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaHandlers/MetaDocs.cs]: What is this inheriting from?

File: SharpDenizenTools/MetaHandlers/MetaDocs.cs [hunk 11]
```diff
@@ -20,76 +13,67 @@ public class MetaDocs
         public static MetaDocs CurrentMeta = null;
 
         /// <summary>The "command" meta type.</summary>
-        public static MetaType META_TYPE_COMMAND = new() { Name = "Command", WebPath = "Commands" };
+        public MetaType<MetaCommand> META_TYPE_COMMAND;
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaHandlers/MetaDocs.cs]: This being changed to no longer static is a potential issue

File: SharpDenizenTools/MetaHandlers/MetaDocs.cs [hunk 12]
```diff
@@ -115,6 +99,23 @@ public class MetaDocs
...
+            META_TYPE_COMMAND = new(this, "Command", "Commands", 512, () => new MetaCommand());
+            META_TYPE_MECHANISM = new(this, "Mechanism", "Mechanisms", 1024, () => new MetaMechanism());
+            META_TYPE_EVENT = new(this, "Event", "Events", 1024, () => new MetaEvent());
+            META_TYPE_ACTION = new(this, "Action", "Actions", 512, () => new MetaAction());
+            META_TYPE_LANGUAGE = new(this, "Language", "Languages", 512, () => new MetaLanguage());
+            META_TYPE_TAG = new(this, "Tag", "Tags", 2048, () => new MetaTag());
+            META_TYPE_OBJECT = new(this, "ObjectType", "ObjectTypes", 512, () => new MetaObjectType());
+            META_TYPE_PROPERTY = new(this, "Property", "Properties", 512, () => new MetaProperty());
+            META_TYPE_GUIDEPAGE = new(this, "GuidePage", null, 512, () => new MetaGuidePage());
+            META_TYPE_DATA = new(this, "Data", null, 0, () => new MetaDataValue());
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaHandlers/MetaDocs.cs]: Internal Data being treated as its own type is an issue, it's intentionally excluded from most systems

File: SharpDenizenTools/MetaHandlers/MetaDocs.cs [hunk 13]
```diff
@@ -239,7 +205,7 @@ public IEnumerable<MetaObject> AllMetaObjects()
         public MetaTag FindTag(string tagText)
         {
             string cleaned = MetaTag.CleanTag(tagText).ToLowerFast();
-            if (Tags.TryGetValue(cleaned, out MetaTag result))
+            if (META_TYPE_TAG.Meta.TryGetValue(cleaned, out MetaTag result))
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaHandlers/MetaDocs.cs]: All of these changes are awkward and redundant

File: SharpDenizenTools/MetaHandlers/MetaDocs.cs [hunk 14]
```diff
@@ -115,6 +98,23 @@ public class MetaDocs
         /// <summary>Set of raw adjustable keys.</summary>
         public HashSet<string> RawAdjustables = [];
 
+        /// <summary>Creates a new instance of <see cref="MetaDocs"/> and registers it's <see cref="MetaTypeData{T}"/>.</summary>
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaHandlers/MetaDocs.cs]: its

Comment (tal5) [reply SharpDenizenTools/MetaHandlers/MetaDocs.cs]: Fixed

File: SharpDenizenTools/MetaObjects/MetaAction.cs [hunk 15]
```diff
@@ -44,6 +35,12 @@ public override void AddTo(MetaDocs docs)
         /// <summary>Determination options. One Determination per string.</summary>
         public string[] Determinations = [];
 
+        /// <inheritdoc/>
+        public override void AddTo(MetaDocs docs)
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaAction.cs]: why was this moved

File: SharpDenizenTools/MetaObjects/MetaType.cs [hunk 16]
```diff
@@ -1,8 +1,4 @@
-﻿using System;
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaType.cs]: This change should probably be removed from the PR

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        /// <summary>Creates a new <see cref="MetaObject"/> using <see cref="Getter"/> and sets it's type.</summary>
+        public MetaObject CreateNewMeta()
+        {
+            MetaObject newMeta = Getter();
+            newMeta.Type = Type;
+            return newMeta;
+        }
+    }
+
+    /// <summary>Data for a specific <see cref="MetaType"/></summary>
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaTypeData.cs]: missing `.` at end

File: SharpDenizenTools/MetaHandlers/MetaDocs.cs [hunk 18]
```diff
@@ -115,6 +98,23 @@ public class MetaDocs
         /// <summary>Set of raw adjustable keys.</summary>
         public HashSet<string> RawAdjustables = [];
 
+        /// <summary>Creates a new instance of <see cref="MetaDocs"/> and registers its <see cref="MetaTypeData{T}"/>.</summary>
+        public MetaDocs()
+        {
+            // Extensions explicitly first
+            IMetaTypeData.Register(this, META_TYPE_EXTENSION, Extensions, () => new MetaExtension());
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaHandlers/MetaDocs.cs]: Why is this a static method on the interface when its purpose is affecting instance data on the MetaDocs instance?

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+namespace SharpDenizenTools.MetaObjects
+{
+    /// <summary>A generic-less <see cref="MetaTypeData{T}"/> interface.</summary>
+    public interface IMetaTypeData
+    {
+        /// <summary>Getter to create a new instance of this meta type.</summary>
+        Func<MetaObject> Getter { get; }
+
+        /// <summary>All meta docs of this type, by name.</summary>
+        Dictionary<string, MetaObject> Meta {  get; }
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaTypeData.cs]: stray double space here

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+        public MetaObject CreateNewMeta()
+        {
+            MetaObject newMeta = Getter();
+            newMeta.Type = Type;
+            return newMeta;
+        }
+    }
+
+    /// <summary>Data for a specific <see cref="MetaType"/>.</summary>
+    public record MetaTypeData<T>(Dictionary<string, T> Meta, Func<T> Getter, MetaType Type) : IMetaTypeData where T : MetaObject
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaTypeData.cs]: `Meta` and `Meta` have the same name on the same instance but different fundamental properties.

The one on the interface should probably be `IReadOnlyDictionary` given that it's... a perpetually generated duplicate of the real data. Ow. maybe should have an `IEnumerable<MetaObject> AllObjects` instead as that's version that's needed and less cursed about it

File: SharpDenizenTools/MetaHandlers/MetaDocs.cs (hunk 18 above)
Comment (tal5) [reply SharpDenizenTools/MetaHandlers/MetaDocs.cs]: Moved to `MetaDocs`

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs (hunk 20 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaTypeData.cs]: Replaced with methods on the interface to directly get meta values, lmk if the naming is fine

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+
+        /// <summary>Gets a <see cref="MetaObject"/> by name, or <c>null</c> if one by that name doesn't exist.</summary>
+        MetaObject MetaByNameIfPresent(string name);
+
+        /// <summary>Returns all <see cref="MetaObject"/>s of this type.</summary>
+        IEnumerable<MetaObject> AllMetaObjects();
+    }
+
+    /// <summary>Data for a specific <see cref="MetaType"/>.</summary>
+    public record MetaTypeData<T>(Dictionary<string, T> Meta, Func<T> Getter, MetaType Type) : IMetaTypeData where T : MetaObject
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaTypeData.cs]: you can actually just require `new()` in the `where` statement and avoid the `Getter` here

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SharpDenizenTools.MetaObjects
+{
+    /// <summary>A generic-less <see cref="MetaTypeData{T}"/> interface.</summary>
+    public interface IMetaTypeData
+    {
+        /// <summary>Getter to create a new instance of this meta type.</summary>
+        Func<MetaObject> Getter { get; }
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaTypeData.cs]: This is a getter-getter, it should probably just be a singular getter

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+        /// <summary>Creates a new <see cref="MetaObject"/> using <see cref="Getter"/> and sets it's type.</summary>
+        public MetaObject CreateNewMeta()
+        {
+            MetaObject newMeta = Getter();
+            newMeta.Type = Type;
+            return newMeta;
+        }
+
+        /// <summary>Gets a <see cref="MetaObject"/> by name, or <c>null</c> if one by that name doesn't exist.</summary>
+        MetaObject MetaByNameIfPresent(string name);
```
Comment (mcmonkey4eva) [inline SharpDenizenTools/MetaObjects/MetaTypeData.cs]: Ideally use the `TryGet` pattern, ie `bool TryGetObject(string name, out MetaObject obj)`

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs (hunk 22 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaTypeData.cs]: Removed in favor of implementing `CreateNewMeta` in `MetaTypeData` (instead of the interface)

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs (hunk 21 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaTypeData.cs]: Done

File: SharpDenizenTools/MetaObjects/MetaTypeData.cs (hunk 23 above)
Comment (tal5) [reply SharpDenizenTools/MetaObjects/MetaTypeData.cs]: Changed to that
